src/gui/sdlmain.cpp

```diff
@@ -2190,6 +2210,57 @@ void GFX_HandleVideoResize(int width, int height)
 	sdl.resizing_window = false;
 }
 
+/* This function is triggered after window is shown to fixup sdl.window
+ * properties in predictable manner on all platforms.
+ *
+ * In specific usecases, certain sdl.window properties might be left unitialized
+ * when starting in fullscreen, which might trigger severe problems for end
+ * users (e.g. placing window partially off-screen, or using fullscreen
+ * resolution for window size).
+ */
+static void FinalizeWindowState()
+{
+	assert(sdl.window);
+
+	if (!sdl.desktop.lazy_init_window_size)
+		return;
+
+	// Don't change window position or size when state changed to
+	// fullscreen.
+	if (sdl.desktop.fullscreen)
+		return;
+
+	// Once window is fixed up once, OS or Window Manager will deal with it
+	// on future expose events.
+	sdl.desktop.lazy_init_window_size = false;
+
+	int w, h;
+	if (sdl.desktop.window.use_original_size) {
+		w = sdl.draw.width * sdl.draw.scalex;
+		h = sdl.draw.height * sdl.draw.scaley;
+	} else {
+		w = sdl.desktop.window.width;
+		h = sdl.desktop.window.height;
+	}
```

Comment:
I think this logic can go up into `SetupWindowResolution`, and we should eliminate all this special handling for `original` window-sizes, and just work with dimensions; `original`, `desktop`, or other human-terms can live at the interface to the config-file, and be processed down into Window dimension.

Reply:
Edit: from the previous comments, `sdl.desktop.window.use_original_size` will remain - but can we also set `sdl.desktop.window.<dims>` to correct values?

For example, in `SetupWindowResolution()`, if `"original"` is set, then we would set `sdl.desktop.window.<dims>` to the splash image dimensions. We could get rid of this branch here.

In `SDL_SetWindowSize` we could also save the applied dimensions to `sdl.desktop.window.<dims>`, so it at least hold correct state information.  All of the critical "original" logic would remain as-is.

Reply:
About storing a copy of *up-to-date* window dimensions in `sdl.desktop.window.*` properties - I think it's not going to work; we can keep a copy in there and update it on the best-known basis, but eventually, it'll not be in sync with whatever OS/WM decided to do with the window at the moment, some examples:
- placing window tiled to the screen edge "half-maximized" (Win 10, Gnome 3)
- "quarter-maximized"(Windows 10)
- using some user-specific window size (any tiling WM)
- moving window between HiDPI and non-HiDPI displays (Win 10, macOS, Wayland in the future), etc, etc.

However, SDL2 does have this information - I don't know how up-to-date it is, but I presume it will be better than whatever hacks we're going to inject to keep it up-to-date in all weird usecases we would need to handle. So when needed, call `SDL_GetWindowSize(sdl.window, &w, &h);` :) Unless you tried it already and it does not work for some reason?

Reply:
I wonder if those window manager adjustments would triggers resize/exposed/.. events? We could update (or invalidate) the window dimensions then.  

Yeah, ultimately I'm hoping we can keep the local sdl.desktop.window struct valid so it can be be relied upon (if not.. maybe we toss the dimension variables and query SDL directly whenever we need them). 

Thanks for exploring this @dreamer; think we've got options going forward.  I'm going to pause effort on instant launch and the toggle fix; I think they will be must easier post 0.75 after disentangling sdlmain!

Reply:
> I wonder if those window manager adjustments would triggers resize/exposed/.. events? We could update (or invalidate) the window dimensions then.

We don't know until we'll test :) Right now, I am 100% sure that DEs using Mutter-derived WMs send us Expose events reliably (whenever window is shown, switched to fullscreen or away from fullscreen). KDE Plasma and MATE do not (sometimes Expose event is missing or arrives much later - e.g. sometimes at the same time as keyboard focus event).  Initially I had "lazy init" part handled on expose, but moving it to "size changed" event made it behave much more reliably.

Resize events arrive differently on different OSes (but it's not SDL2 fault! SDL is pumping these events in, on Linux they arrive correctly, on non-Linux systems they are blocked somewhere on OS thread handling :/). Fortunately, they can be received via callback off-main-thread, and that makes the order of events somewhat consistent (this will be in next PR). From what I've googled, it is a common problem to game developers out there.